File: schemas/loyalty-rewards-program/convex/tiers.ts
// tiers: Tier/VIP level definitions with qualification thresholds and ordering.
// See README.md for full design rationale.
import { defineTable } from "convex/server";
import { v } from "convex/values";

export const tiers = defineTable({
  programId: v.id("loyalty_programs"),
  name: v.string(),
  slug: v.string(),
  description: v.optional(v.string()),
  position: v.number(),
  qualificationType: v.union(
    v.literal("points_earned"),
    v.literal("amount_spent"),
    v.literal("transaction_count")
  ),
  qualificationValue: v.number(),
  qualificationPeriodDays: v.optional(v.number()),
  retainDays: v.optional(v.number()),
  iconUrl: v.optional(v.string()),
  color: v.optional(v.string()),
  isDefault: v.boolean(),
  metadata: v.optional(v.any()),
  updatedAt: v.number(),
})
  .index("by_program_id_and_slug", ["programId", "slug"])
  .index("by_program_id_and_position", ["programId", "position"])
  .index("by_is_default", ["isDefault"]);

File: schemas/loyalty-rewards-program/spacetimedb/tiers.rs
// tiers: Tier/VIP level definitions with qualification thresholds and ordering.
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum TierQualificationType {
    PointsEarned,      // type: String
    AmountSpent,
    TransactionCount,
}

#[spacetimedb::table(name = tiers, public)]
pub struct Tier {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub program_id: String, // UUID, FK → loyalty_programs.id (cascade delete)
    // Composite unique: (program_id, slug) — enforce in reducer logic
    // Composite unique: (program_id, position) — enforce in reducer logic
    pub name: String,
    pub slug: String,
    pub description: Option<String>,
    pub position: i32,
    pub qualification_type: TierQualificationType,
    pub qualification_value: i32,
    pub qualification_period_days: Option<i32>,
    pub retain_days: Option<i32>,
    pub icon_url: Option<String>,
    pub color: Option<String>,
    #[index(btree)]
    pub is_default: bool, // default false
    pub metadata: Option<String>, // JSON stored as string
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
